pysaliency/numba_utils.py

```diff
@@ -47,3 +47,43 @@ def _auc_for_one_positive(positive, negatives):
             count += 0.5
 
     return count / len(negatives)
+
+
+def general_roc_numba(positives, negatives, judd=0):
+    sorted_positives = np.sort(positives)[::-1]
+    sorted_negatives = np.sort(negatives)[::-1]
+
+    if judd == 0:
+        all_values = np.hstack([positives, negatives])
+        all_values = np.sort(all_values)[::-1]
+    else:
+        min_val = min(sorted_positives[len(positives)-1], sorted_negatives[len(negatives)-1])
+        max_val = max(sorted_positives[0], sorted_negatives[0]) + 1
+        all_values = np.hstack((max_val, positives, min_val))
+        all_values = np.sort(all_values)[::-1]
+
+    false_positive_rates = np.zeros(len(all_values) + 1)
+    hit_rates = np.zeros(len(all_values) + 1)
+    positive_count = len(positives)
+    negative_count = len(negatives)
+    return _general_roc_numba(all_values, sorted_positives, sorted_negatives, positive_count, negative_count, false_positive_rates, hit_rates)
+
+
+@numba.jit(nopython=True)
+def _general_roc_numba(all_values, sorted_positives, sorted_negatives, positive_count, negative_count, false_positive_rates, hit_rates):
+    """calculate ROC score for given values of positive and negative
+    distribution"""
+
+    true_positive_count = 0
+    false_positive_count = 0
+    for i in range(len(all_values)):
+        theta = all_values[i]
+        while true_positive_count < positive_count and sorted_positives[true_positive_count] >= theta:
+            true_positive_count += 1
+        while false_positive_count < negative_count and sorted_negatives[false_positive_count] >= theta:
+            false_positive_count += 1
+        false_positive_rates[i+1] = float(false_positive_count) / negative_count
+        hit_rates[i+1] = float(true_positive_count) / positive_count
+
+    auc = np.trapz(hit_rates, false_positive_rates)
```

Comment:
does this numpy call work with `nopython=True`? Anyway, I would suggest to call it in the `general_roc_numba` function and have `_general_roc_numba` only compute the hit rates and false positive rate

Reply:
yes this numpy call does seem to work. But we need the hit_rates and false_positive_rates to calculate the auc and those are being calculated in the  _general_roc_numba.